### ISSUE TITLE
Consistant crash and memory problems with reductions on MI100 and MI250

### ISSUE THREAD

**Issue by Keluaa** [JuliaGPU/AMDGPU.jl#338]
Because the all versions of AMDGPU.jl that I tried don't recognize the MI250 GPU (should I make a separate issue about this, or is this known?), I switched to using [Julia v1.9.0-alpha1](https://github.com/JuliaLang/julia/tree/release-1.9), since it uses LLVM 14 which is also the version which is used by the ROCm stack on the cluster. In this case all GPUs on the node are correctly identified.

While AMDGPU.jl can properly work with the MI250, it irregularly fails (all tests are done using only the `-O3` option):

```julia
julia> using AMDGPU

julia> using BenchmarkTools

julia> A = ROCArray(rand(Float64, 1000000));

julia> reduce(+, A)
LoadLib(libhsa-amd-aqlprofile64.so) failed: libhsa-amd-aqlprofile64.so: cannot open shared object file: No such file or directory
500233.3343854438

julia> reduce(+, A) == reduce(+, Array(A))

julia> @benchmark reduce(+, A)
Warning: (false && "Can't find address in allocation map") in hsa_status_t rocr::core::Runtime::FreeMemory(void *), /workspace/srcdir/ROCR-Runtime-rocm-5.2.3/src/core/runtime/runtime.cpp:304
Warning: (false && "Can't find address in allocation map") in hsa_status_t rocr::core::Runtime::FreeMemory(void *), /workspace/srcdir/ROCR-Runtime-rocm-5.2.3/src/core/runtime/runtime.cpp:304
BenchmarkTools.Trial: 1986 samples with 1 evaluation.
 Range (min … max):  1.174 ms …   1.954 s  ┊ GC (min … max): 0.00% … 0.56%
 Time  (median):     1.545 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.510 ms ± 43.810 ms  ┊ GC (mean ± σ):  0.22% ± 0.01%

                        ▅▇█               ▁        ▁▂▁        
  ▂▂▃▂▂▂▂▁▂▂▁▁▁▁▂▂▁▁▃▄▆█████▄▃▂▂▂▂▄▅▆▆▆▆▆▆██▇▆▆▆▅▆▅████▇▆▇▆▄ ▄
  1.17 ms        Histogram: frequency by time        1.71 ms <

 Memory estimate: 22.25 KiB, allocs estimate: 778.

julia> julia> @benchmark reduce(+, A)

[305753] signal (11.1): Segmentation fault
```

Here I used BenchmarkTools to call `reduce` repeatedly, but I didn't manage to cause the bug by calling it in a loop (~10000 iterations).
I can try the same thing in 2 or 3 different sessions without triggering the bug.

<details>
  <summary>Here is the full stacktrace</summary>

```
in expression starting at REPL[7]:1
_ZNK4rocr4core5CheckILm8213662378061812342ELb1EE7IsValidEv at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZNK4rocr4core12SharedSignal7IsValidEv at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZN4rocr3HSA18hsa_signal_destroyE12hsa_signal_s at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
hsa_signal_destroy at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
hsa_signal_destroy at /work/.julia/packages/AMDGPU/FA0AP/src/hsa/LibHSARuntime.jl:253
unknown function (ip: 0x145e69d61781)
#32 at /work/.julia/packages/AMDGPU/FA0AP/src/runtime/signal.jl:38
unknown function (ip: 0x145e69d616d5)
run_finalizer at /work/julia-llvm14-1.9/src/gc.c:293
jl_gc_run_finalizers_in_list at /work/julia-llvm14-1.9/src/gc.c:383
run_finalizers at /work/julia-llvm14-1.9/src/gc.c:429
run_finalizers at /work/julia-llvm14-1.9/src/gc.c:410 [inlined]
ijl_gc_collect at /work/julia-llvm14-1.9/src/gc.c:3585
gc at ./gcutils.jl:98 [inlined]
gc at ./gcutils.jl:98 [inlined]
gcscrub at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:4 [inlined]
#_run#48 at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:95
_run at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:92
unknown function (ip: 0x145e69d700f6)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
jl_f__call_latest at /work/julia-llvm14-1.9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
#run_result#45 at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:34 [inlined]
run_result at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:34 [inlined]
#run#49 at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:117
run at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:117 [inlined]
run at /work/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:117
unknown function (ip: 0x145e69d6f932)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
do_call at /work/julia-llvm14-1.9/src/interpreter.c:126
eval_value at /work/julia-llvm14-1.9/src/interpreter.c:226
eval_stmt_value at /work/julia-llvm14-1.9/src/interpreter.c:177 [inlined]
eval_body at /work/julia-llvm14-1.9/src/interpreter.c:606
jl_interpret_toplevel_thunk at /work/julia-llvm14-1.9/src/interpreter.c:762
jl_toplevel_eval_flex at /work/julia-llvm14-1.9/src/toplevel.c:912
jl_toplevel_eval_flex at /work/julia-llvm14-1.9/src/toplevel.c:856
ijl_toplevel_eval_in at /work/julia-llvm14-1.9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:152
repl_backend_loop at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:248
#start_repl_backend#46 at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:233
start_repl_backend at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:230 [inlined]
#run_repl#59 at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:376
run_repl at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:362
jfptr_run_repl_58026 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
#1016 at ./client.jl:421
jfptr_YY.1016_31981 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
jl_f__call_latest at /work/julia-llvm14-1.9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_49673 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
true_main at /work/julia-llvm14-1.9/src/jlapi.c:573
jl_repl_entrypoint at /work/julia-llvm14-1.9/src/jlapi.c:717
main at /work/julia-llvm14-1.9/cli/loader_exe.c:59
__libc_start_main at /lib64/libc.so.6 (unknown line)
_start at /work/julia-llvm14-1.9/julia (unknown line)
Allocations: 59687122 (Pool: 59656792; Big: 30330); GC: 80
Segmentation fault (core dumped)
```
  
</details>

On MI100, I can also sometimes reproduce this crash.

For a more complex example, I have a consistent crash:
```julia
julia> using AMDGPU

julia> N = 10000000;

julia> A = rand(Float64, N); Ad = ROCArray(A);

julia> B = rand(Float64, N); Bd = ROCArray(B);

julia> M = ones(Float64, N) .* (rand(Float64, N) .> 0.5); Md = ROCArray(M);

julia> r = 50:N-50;

julia> for i in 1:100
           isapprox(reduce(min, @views (A[r] .* B[r] .* M[r])), reduce(min, @views (Ad[r] .* Bd[r] .* Md[r])))
           println("$i")
       end
LoadLib(libhsa-amd-aqlprofile64.so) failed: libhsa-amd-aqlprofile64.so: cannot open shared object file: No such file or directory
1
2
3
4
5
6
7
8
9
julia: /workspace/srcdir/ROCR-Runtime-rocm-5.2.3/src/core/runtime/signal.cpp:134: void rocr::core::Signal::registerIpc(): Assertion `ipcMap_.find(handle.handle) == ipcMap_.end() && "Can't register the same IPC signal twice."' failed.

[1812412] signal (6.-6): Aborted
``` 

<details>
  <summary>Here is the full stacktrace</summary>

```
julia: /workspace/srcdir/ROCR-Runtime-rocm-5.2.3/src/core/runtime/signal.cpp:134: void rocr::core::Signal::registerIpc(): Assertion `ipcMap_.find(handle.handle) == ipcMap_.end() && "Can't register the same IPC signal twice."' failed.

[1812412] signal (6.-6): Aborted
in expression starting at REPL[8]:1
gsignal at /lib64/libc.so.6 (unknown line)
abort at /lib64/libc.so.6 (unknown line)
__assert_fail_base.cold.0 at /lib64/libc.so.6 (unknown line)
__assert_fail at /lib64/libc.so.6 (unknown line)
_ZN4rocr4core6Signal11registerIpcEv at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZN4rocr4core6SignalC2EPNS0_12SharedSignalEb at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZN4rocr4core14BusyWaitSignalC1EPNS0_12SharedSignalEb at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZN4rocr4core13DefaultSignalC2Elb at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
_ZN4rocr3AMD21hsa_amd_signal_createEljPK11hsa_agent_smP12hsa_signal_s at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
hsa_amd_signal_create at /work/.julia/artifacts/edcd56051c809324ebfbf4068ec18f917e13ac52/lib/libhsa-runtime64.so (unknown line)
hsa_amd_signal_create at /work/.julia/packages/AMDGPU/FA0AP/src/hsa/LibHSARuntime.jl:1561
unknown function (ip: 0x145795705100)
#ROCSignal#31 at /work/.julia/packages/AMDGPU/FA0AP/src/runtime/signal.jl:30
ROCSignal at /work/.julia/packages/AMDGPU/FA0AP/src/runtime/signal.jl:24 [inlined]
ROCSignal at /work/.julia/packages/AMDGPU/FA0AP/src/runtime/signal.jl:24 [inlined]
create_event at /work/.julia/packages/AMDGPU/FA0AP/src/highlevel.jl:124 [inlined]
macro expansion at /work/.julia/packages/AMDGPU/FA0AP/src/highlevel.jl:440 [inlined]
#mapreducedim!#78 at /work/.julia/packages/AMDGPU/FA0AP/src/mapreduce.jl:186
mapreducedim! at /work/.julia/packages/AMDGPU/FA0AP/src/mapreduce.jl:87 [inlined]
#_mapreduce#22 at /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:69
_mapreduce at /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:35
unknown function (ip: 0x145795718871)
#mapreduce#20 at /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:31
mapreduce at /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:31 [inlined]
#reduce#804 at ./reducedim.jl:406 [inlined]
reduce at ./reducedim.jl:406
unknown function (ip: 0x14579570e186)
top-level scope at ./REPL[8]:2
jl_toplevel_eval_flex at /work/julia-llvm14-1.9/src/toplevel.c:903
jl_toplevel_eval_flex at /work/julia-llvm14-1.9/src/toplevel.c:856
ijl_toplevel_eval_in at /work/julia-llvm14-1.9/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
eval_user_input at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:152
repl_backend_loop at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:248
#start_repl_backend#46 at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:233
start_repl_backend at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:230 [inlined]
#run_repl#59 at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:376
run_repl at /work/julia-llvm14-1.9/usr/share/julia/stdlib/v1.9/REPL/src/REPL.jl:362
jfptr_run_repl_58026 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
#1016 at ./client.jl:421
jfptr_YY.1016_31981 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
jl_f__call_latest at /work/julia-llvm14-1.9/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
run_main_repl at ./client.jl:405
exec_options at ./client.jl:322
_start at ./client.jl:522
jfptr__start_49673 at /work/julia-llvm14-1.9/usr/lib/julia/sys.so (unknown line)
jl_apply at /work/julia-llvm14-1.9/src/julia.h:1868 [inlined]
true_main at /work/julia-llvm14-1.9/src/jlapi.c:573
jl_repl_entrypoint at /work/julia-llvm14-1.9/src/jlapi.c:717
main at /work/julia-llvm14-1.9/cli/loader_exe.c:59
__libc_start_main at /lib64/libc.so.6 (unknown line)
_start at /work/julia-llvm14-1.9/julia (unknown line)
Allocations: 48028281 (Pool: 47997884; Big: 30397); GC: 67
Aborted (core dumped)
```
  
</details>

The crash always occurs at the 10th iteration, both on MI100 and MI250, and doesn't happen on Julia 1.7.2.

On a bigger code which uses a dozen of arrays of 10^8 doubles, I get a consistant error at the same point of the solver, which uses a reduction: HSA_STATUS_ERROR_OUT_OF_RESOURCES.

<details>
  <summary>Full message</summary>

```julia
Failed to map requested scratch (3520331776) - reducing queue occupancy.
  1760165888 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3523739648) - reducing queue occupancy.
  1761869824 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3523739648) - reducing queue occupancy.
  1761869824 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3510108160) - reducing queue occupancy.
  1755054080 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3520331776) - reducing queue occupancy.
  1760165888 scratch mapped, 50.00% occupancy.
Failed to map requested scratch (3520331776) - reducing queue occupancy.
  Could not allocate scratch for one wave per CU.
^CERROR: QueueError(Queue at Ptr{AMDGPU.HSA.LibHSARuntime.hsa_queue_s} @0x000014678c052000) due to:
HSA error (code #4104, HSA_STATUS_ERROR_OUT_OF_RESOURCES: The runtime failed to allocate the necessary resources. This error may also occur when the core runtime library needs to spawn threads or create internal OS-specific events.)
Stacktrace:
  [1] cleanup!(kersig::AMDGPU.Runtime.ROCKernelSignal; finished::Bool, check_exceptions::Bool)
    @ AMDGPU.Runtime /work/.julia/packages/AMDGPU/FA0AP/src/runtime/kernel-signal.jl:80
  [2] cleanup!
    @ /work/.julia/packages/AMDGPU/FA0AP/src/runtime/kernel-signal.jl:56 [inlined]
  [3] wait(kersig::AMDGPU.Runtime.ROCKernelSignal; check_exceptions::Bool, signal_kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ AMDGPU.Runtime /work/.julia/packages/AMDGPU/FA0AP/src/runtime/kernel-signal.jl:52
  [4] wait
    @ /work/.julia/packages/AMDGPU/FA0AP/src/runtime/kernel-signal.jl:34 [inlined]
  [5] mapreducedim!(f::typeof(identity), op::typeof(+), R::AMDGPU.ROCVector{Float64}, A::Base.Broadcast.Broadcasted{AMDGPU.ROCArrayStyle{1}, Tuple{Base.OneTo{Int64}}, typeof(identity), Tuple{AMDGPU.ROCVector{Float64}}}; init::Float64)
    @ AMDGPU /work/.julia/packages/AMDGPU/FA0AP/src/mapreduce.jl:186
  [6] mapreducedim!
    @ /work/.julia/packages/AMDGPU/FA0AP/src/mapreduce.jl:87 [inlined]
  [7] _mapreduce(f::typeof(identity), op::typeof(+), As::AMDGPU.ROCVector{Float64}; dims::Colon, init::Nothing)
    @ GPUArrays /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:69
  [8] mapreduce(::Function, ::Function, ::AMDGPU.ROCVector{Float64}; dims::Function, init::Nothing)
    @ GPUArrays /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:31
  [9] mapreduce
    @ /work/.julia/packages/GPUArrays/fqD8z/src/host/mapreduce.jl:31 [inlined]
 [10] #reduce#804
    @ ./reducedim.jl:406 [inlined]
 [11] reduce(op::Function, A::AMDGPU.ROCVector{Float64})
    @ Base ./reducedim.jl:406
```
  
</details>

Julia hangs after the "Could not allocate scratch for one wave per CU." message, it is only after pressing Ctrl+C that it prints the error message. 
Sadly I did not managed to boil it down to a simple case, but the two previous crashes happen for a very similar line of code.

**Comment by Keluaa** [comment]
Upon further testing, the second example consistantly crashes with Julia 1.7.2, but the crash happens around the 30th iteration.
The stacktrace is exactly the same.

**Comment by pxl-th** [comment]
> I switched to using [Julia v1.9.0-alpha1](https://github.com/JuliaLang/julia/tree/release-1.9), since it uses LLVM 14

Yeah, you should use Julia with the matching LLVM version.

> Warning: (false && "Can't find address in allocation map") in hsa_status_t rocr::core::Runtime::FreeMemory(void *), /workspace/srcdir/ROCR-Runtime-rocm-5.2.3/src/core/runtime/runtime.cpp:304

I think this was fixed on `jpd/dev` branch, you can try it out along with `GPUCompiler.jl#vc/always-inline` branch.

> Failed to map requested scratch (3520331776) - reducing queue occupancy.
  1760165888 scratch mapped, 50.00% occupancy.

I'm very curious why this happens, as I have it too with my kernels, which I'm currently trying to rewrite to avoid it.
One guess was that functions in kernels are not inlined, for which you should try above branches.

> ROCm stack on the cluster

I think AMDGPU.jl currently uses its own artifacts for ROCm stack, not the ones on your system, judging by the:
`LoadLib(libhsa-amd-aqlprofile64.so) failed: libhsa-amd-aqlprofile64.so: cannot open shared object file: No such file or directory`.
You can force it to use system ones by setting `JULIA_AMDGPU_DISABLE_ARTIFACTS=1` env variable.

**Comment by Keluaa** [comment]
> I think this was fixed on `jps/dev` branch, you can try it out along with `GPUCompiler.jl#vc/always-inline` branch.

I tried it on 85a0b147cf7afe7a80eaed7e067d42370cb71abf (the lastest commits of `jps/dev` had some bugs), and the message indeed disappeared.

> I think AMDGPU.jl currently uses its own artifacts for ROCm stack

Yes, for some reasons my the ROCm stask of my system causes more problems than the one packaged with AMDGPU.jl, so that is why I am working with the artifacts. It is by trying to use the ROCm stask of my system that I found out about the matching LLVM versions thing.

**Comment by pxl-th** [comment]
RE

> Failed to map requested scratch (3520331776) - reducing queue occupancy.
1760165888 scratch mapped, 50.00% occupancy.

My guess that in AMDGPU is not keeping up with freeing the resources.
For example, you can get: `HSA exception: MemoryRegion::BlockAllocator::alloc failed.`

Simply by running:
```julia
julia> for _ in 1:10000
           ROCArray{UInt32}(undef, 128^3)
       end
```

Which is often appears right before `reduced queue occupancy`.

**Comment by pxl-th** [comment]
So if it's allocation for the kernel you get `reduced queue occupancy`;
same for regular memory buffer but with `HSA exception: MemoryRegion::BlockAllocator::alloc failed.`

**Comment by pxl-th** [comment]
Calling `GC.gc()` at each loop iteration solves the issue for me.

**Comment by Keluaa** [comment]
Calling `GC.gc()` is definitely the best way to circumvent the bug.

From my testing, you don't even need to call it at each iteration to prevent the crash in my second example: simply calling it once after JIT compilation causes no problems.

```julia
julia> using AMDGPU

julia> N = 10000000;

julia> A = rand(Float64, N); Ad = ROCArray(A);

julia> B = rand(Float64, N); Bd = ROCArray(B);

julia> M = ones(Float64, N) .* (rand(Float64, N) .> 0.5); Md = ROCArray(M);

julia> r = 50:N-50;

julia> isapprox(reduce(min, @views (A[r] .* B[r] .* M[r])), reduce(min, @views (Ad[r] .* Bd[r] .* Md[r])))

julia> GC.gc()

julia> for i in 1:1000
           isapprox(reduce(min, @views (A[r] .* B[r] .* M[r])), reduce(min, @views (Ad[r] .* Bd[r] .* Md[r])))
       end
```

On the bigger code I am developing, running the garbage collector once after compilation and then once every 5 or 10 iterations is enough to get rid of all the problems, on Julia 1.7.2 and Julia 1.9.0-alpha1.

**Comment by jpsamaroo** [comment]
@Keluaa thanks for the report and reproducer codes! @pxl-th has been looking into these issues for the past few weeks, and is making some progress on locating the root cause. With the holidays being upon us, progress on this might slow a little bit, but this is currently our top-priority bug for AMDGPU.jl.

**Comment by pxl-th** [comment]
This should no longer happen.
This was most likely the same issue as in: https://github.com/JuliaGPU/AMDGPU.jl/issues/388#issuecomment-1625458560
Feel free to reopen though.
